File: AudioApp/backend/models/track.py
"""
Track model for music library management.
"""

from sqlalchemy import Column, Integer, String, Float, DateTime, Boolean
from sqlalchemy.ext.declarative import declarative_base
from datetime import datetime
from pydantic import BaseModel
from typing import Optional

Base = declarative_base()


class TrackDB(Base):
    """SQLAlchemy model for music tracks."""

    __tablename__ = "tracks"

    id = Column(Integer, primary_key=True, index=True)
    file_path = Column(String, unique=True, nullable=False, index=True)
    title = Column(String, index=True)
    artist = Column(String, index=True)
    album = Column(String, index=True)
    genre = Column(String)
    year = Column(Integer)
    duration = Column(Float)  # in seconds
    track_number = Column(Integer)
    bitrate = Column(Integer)
    sample_rate = Column(Integer)
    file_size = Column(Integer)
    date_added = Column(DateTime, default=datetime.utcnow)
    last_played = Column(DateTime, nullable=True)
    play_count = Column(Integer, default=0)
    favorite = Column(Boolean, default=False)

    def to_dict(self):
        """Convert track to dictionary."""
        return {
            "id": self.id,
            "file_path": self.file_path,
            "title": self.title,
            "artist": self.artist,
            "album": self.album,
            "genre": self.genre,
            "year": self.year,
            "duration": self.duration,
            "track_number": self.track_number,
            "bitrate": self.bitrate,
            "sample_rate": self.sample_rate,
            "file_size": self.file_size,
            "date_added": self.date_added.isoformat() if self.date_added else None,
            "last_played": self.last_played.isoformat() if self.last_played else None,
            "play_count": self.play_count,
            "favorite": self.favorite,
        }


class Track(BaseModel):
    """Pydantic model for API requests/responses."""

    id: Optional[int] = None
    file_path: str
    title: Optional[str] = None
    artist: Optional[str] = None
    album: Optional[str] = None
    genre: Optional[str] = None
    year: Optional[int] = None
    duration: Optional[float] = None
    track_number: Optional[int] = None
    bitrate: Optional[int] = None
    sample_rate: Optional[int] = None
    file_size: Optional[int] = None
    date_added: Optional[datetime] = None
    last_played: Optional[datetime] = None
    play_count: int = 0
    favorite: bool = False

    class Config:
        from_attributes = True


class PlaybackState(BaseModel):
    """Current playback state."""

    is_playing: bool = False
    current_track: Optional[Track] = None
    position: float = 0.0  # Current position in seconds
    volume: float = 1.0  # 0.0 to 1.0
    is_shuffled: bool = False
    repeat_mode: str = "none"  # none, one, all

File: AudioApp/AudioApp/Models/Track.swift
//
//  Track.swift
//  AudioApp
//
//  Music track model matching Python backend API
//

import Foundation

struct Track: Codable, Identifiable {
    let id: Int?
    let filePath: String
    let title: String?
    let artist: String?
    let album: String?
    let genre: String?
    let year: Int?
    let duration: Double?
    let trackNumber: Int?
    let bitrate: Int?
    let sampleRate: Int?
    let fileSize: Int?
    let dateAdded: String?
    let lastPlayed: String?
    let playCount: Int
    let favorite: Bool

    enum CodingKeys: String, CodingKey {
        case id
        case filePath = "file_path"
        case title
        case artist
        case album
        case genre
        case year
        case duration
        case trackNumber = "track_number"
        case bitrate
        case sampleRate = "sample_rate"
        case fileSize = "file_size"
        case dateAdded = "date_added"
        case lastPlayed = "last_played"
        case playCount = "play_count"
        case favorite
    }

    var displayTitle: String {
        title ?? "Unknown Track"
    }

    var displayArtist: String {
        artist ?? "Unknown Artist"
    }

    var displayAlbum: String {
        album ?? "Unknown Album"
    }

    var formattedDuration: String {
        guard let duration = duration else { return "--:--" }
        let minutes = Int(duration) / 60
        let seconds = Int(duration) % 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

struct PlaybackState: Codable {
    let isPlaying: Bool
    let isPaused: Bool
    let currentTrack: Track?
    let position: Double
    let volume: Double
    let isShuffled: Bool
    let repeatMode: String
    let queueLength: Int
    let queuePosition: Int

    enum CodingKeys: String, CodingKey {
        case isPlaying = "is_playing"
        case isPaused = "is_paused"
        case currentTrack = "current_track"
        case position
        case volume
        case isShuffled = "is_shuffled"
        case repeatMode = "repeat_mode"
        case queueLength = "queue_length"
        case queuePosition = "queue_position"
    }
}

struct LibraryStats: Codable {
    let totalTracks: Int
    let totalDurationSeconds: Double
    let totalSizeBytes: Int
    let uniqueArtists: Int
    let uniqueAlbums: Int

    enum CodingKeys: String, CodingKey {
        case totalTracks = "total_tracks"
        case totalDurationSeconds = "total_duration_seconds"
        case totalSizeBytes = "total_size_bytes"
        case uniqueArtists = "unique_artists"
        case uniqueAlbums = "unique_albums"
    }

    var formattedDuration: String {
        let hours = Int(totalDurationSeconds) / 3600
        let minutes = (Int(totalDurationSeconds) % 3600) / 60
        return "\(hours)h \(minutes)m"
    }

    var formattedSize: String {
        let gb = Double(totalSizeBytes) / 1_073_741_824
        return String(format: "%.2f GB", gb)
    }
}
